detector_test.go

```diff
@@ -284,486 +698,418 @@ func testDetector(t *testing.T, when spec.G, it spec.S) {
-							{ID: "C", Version: "v2"},
-						},
-						Requires: []lifecycle.Require{{Name: "dep1"}, {Name: "dep1"}},
+			if s := h.AllLogs(logHandler); !strings.HasSuffix(s,
+				"======== Results ========\n"+
+					"pass: A@v1\n"+
+					"pass: C@v2\n"+
+					"pass: D@v2\n"+
+					"pass: B@v1\n"+
+					"Resolving plan... (try #1)\n"+
+					"A v1\n"+
+					"C v2\n"+
+					"D v2\n"+
+					"B v1\n",
+			) {
+				t.Fatalf("Unexpected log:\n%s\n", s)
+			}
+		})
+
+		it("should fail if all requires are not provided first", func() {
+			group := []buildpack.GroupBuildpack{
+				{ID: "A", Version: "v1", Optional: true},
+				{ID: "B", Version: "v1"},
+				{ID: "C", Version: "v1"},
+			}
+
+			detectRuns := &sync.Map{}
+			detectRuns.Store("A@v1", buildpack.DetectRun{
+				PlanSections: buildpack.PlanSections{
+					Provides: []buildpack.Provide{
+						{Name: "dep1"},
 					},
-					{
-						Providers: []lifecycle.GroupBuildpack{
-							{ID: "A", Version: "v1"},
-							{ID: "C", Version: "v2"},
-							{ID: "D", Version: "v2"},
-						},
-						Requires: []lifecycle.Require{{Name: "dep2"}, {Name: "dep2"}, {Name: "dep2"}},
+				},
+				Code: 100,
+			})
+			detectRuns.Store("B@v1", buildpack.DetectRun{
+				PlanSections: buildpack.PlanSections{
+					Requires: []buildpack.Require{
+						{Name: "dep1"},
 					},
-				}) {
-					t.Fatalf("Unexpected entries:\n%+v\n", plan.Entries)
-				}
-
-				if s := allLogs(logHandler); !strings.HasSuffix(s,
-					"======== Results ========\n"+
-						"pass: A@v1\n"+
-						"pass: C@v2\n"+
-						"pass: D@v2\n"+
-						"pass: B@v1\n"+
-						"Resolving plan... (try #1)\n"+
-						"A v1\n"+
-						"C v2\n"+
-						"D v2\n"+
-						"B v1\n",
-				) {
-					t.Fatalf("Unexpected log:\n%s\n", s)
-				}
+				},
+			})
+			detectRuns.Store("C@v1", buildpack.DetectRun{
+				PlanSections: buildpack.PlanSections{
+					Provides: []buildpack.Provide{
+						{Name: "dep1"},
+					},
+					Requires: []buildpack.Require{
+						{Name: "dep1"},
+					},
+				},
 			})
 
-			it("should fail if all requires are not provided first", func() {
-				toappfile("\n[[provides]]\n name = \"dep1\"", "detect-plan-A-v1.toml", "detect-plan-C-v1.toml")
-				toappfile("\n[[requires]]\n name = \"dep1\"", "detect-plan-B-v1.toml", "detect-plan-C-v1.toml")
-				mkappfile("100", "detect-status-A-v1")
-
-				_, _, err := lifecycle.BuildpackOrder{
-					{Group: []lifecycle.GroupBuildpack{
-						{ID: "A", Version: "v1", Optional: true},
-						{ID: "B", Version: "v1"},
-						{ID: "C", Version: "v1"},
-					}},
-				}.Detect(config)
-				if err, ok := err.(*lifecycle.Error); !ok || err.Type != lifecycle.ErrTypeFailedDetection {
-					t.Fatalf("Unexpected error:\n%s\n", err)
-				}
+			//toappfile("\n[[provides]]\n name = \"dep1\"", "detect-plan-A-v1.toml", "detect-plan-C-v1.toml")
+			//toappfile("\n[[requires]]\n name = \"dep1\"", "detect-plan-B-v1.toml", "detect-plan-C-v1.toml")
+			//mkappfile("100", "detect-status-A-v1")
```

Comment:
These tests are fairly complicated, so leaving this comment with the old setup so that we can verify it was translated correctly.